docs: fix missing images etc. (#11)
onomondo/nrf-softsim: README.md

- [Understanding the SIM - why SoftSIM is possible](#understanding-the-sim---why-softsim-is-possible)
- [Details on provisioning](#provisioning)
- [Details on kConfig options](#kconfig-options)

### Setup
For existing toolchains and build systems it is sufficient to update the manifest to point to `west.yml` inside this repository:

```
cd <ncs_base>
git clone git@github.com:onomondo/nrf-softsim.git modules/lib/onomondo-softsim
west config manifest.path modles/lib/onomondo-softsim/
west update
```

First time setting it up? We recommend using the [nRF Connect for Desktop](https://www.nordicsemi.com/Products/Development-tools/nrf-connect-for-desktop) to get the build system correctly set up. Once done, configure the manifest as described above.

You can verify that the SoftSIM code has been cloned by examining the `<ncs_base>/modules/lib/onomondo-softsim` path.

Bonus tip: The Toolchain Manager allows you to easily generate the correct environment variables. Click the small arrow and select `Generate environment script`. The output file contains everything you need to set up the new toolchain. 

### Get access to your free SoftSIM profiles

> SoftSIM profiles are delivered through our API. As this can be a bit cumbersome we've developed a small tool to make this process a bit easier. The tool is available at [sofsim-cli](https://github.com/onomondo/onomondo-softsim-cli). Additional instructions can be found in the cli repository. 

Reviewer: ```suggestion
SoftSIM profiles are delivered through our API. As this can be a bit cumbersome, we've developed a small tool to make this process easier. The tool is available at [sofsim-cli](https://github.com/onomondo/onomondo-softsim-cli). Additional instructions can be found in the CLI repository. 
```

1. Generate an API key on [app.onomondo.com/api-keys](https://app.onomondo.com/api-keys). Follow the instructions on the app. 
2. Download the `softsim` cli tool for your platform.
3. Fetch your profiles: `./softsim fetch --api-key = <your_api_key> -n 5`. This will create a `profiles` directory for you with `5` encrypted profiles.

> Everytime you require a new profiles simple use the `./softsim next --key=<path to your private key>`. It will look in the `./profiles` folder and decrypt and format a profile. _This command guarentees that a new profile is given each time._

Reviewer: ```suggestion
Every time you require a new profile, simply use the `./softsim next --key=<path to your private key>`. It will look in the `./profiles` folder and decrypt and format a profile. _This command guarantees that a new profile is given each time._
```

### Configure and build

> There currently two samples included so showcase how a SoftSIM can be provisioned. It is recommended to start with the *static profile* sample.

Reviewer: ```suggestion
There are currently two samples included to showcase how a SoftSIM can be provisioned. It is recommended to start with the *static profile* sample.
```

##### static profile
`samples/softsim_static_profile` will provision a profile during the first system initialization. The profile is configured in the `prj.conf` 

> In the `prj.conf` you'll find the follwing options related to the SoftSIM statically provisioned profile. This setup is handy for development as the profile doesn't have to be re-provisioned everytime the device is flashed. 

Reviewer: ```suggestion
In the `prj.conf` you'll find the following options related to the SoftSIM statically provisioned profile. This setup is useful for development, as the profile doesn't have to be re-provisioned every time the device is flashed. 
```

```
CONFIG_SOFTSIM_STATIC_PROFILE_ENABLE=y
CONFIG_SOFTSIM_STATIC_PROFILE="011208091..."
```

> Run `./softsim next --key=~/myPrivateKey` (with path to your private key) and grab the output. By default it formats the profile to be accepted by any nrf91 series devices. The profile probably looks something like `01120...`. Replace the `CONFIG_SOFTSIM_STATIC_PROFILE` value with your SoftSIM profile.

Reviewer: ```suggestion
Run `./softsim next --key=~/myPrivateKey` (with path to your private key) and grab the output. By default it formats the profile to be accepted by any nRF91 series devices. The profile will look similar to `01120...`. Replace the `CONFIG_SOFTSIM_STATIC_PROFILE` value with your SoftSIM profile.
```

[00:00:07.096,221] <inf> softsim_sample: Network registration status: Connected - roaming
[00:00:07.096,405] <inf> softsim_sample: LTE connected!
```
### General usage

For most samples and applications, it's sufficient to build by executing the following command:
```
west build -b nrf9160dk_nrf9160_ns -- "-DOVERLAY_CONFIG=$PATH_TO_ONOMONDO_SOFTSIM/overlay-softsim.conf"
```
Where `PATH_TO_ONOMONDO_SOFTSIM` is the path of the downloaded Onomondo SoftSIM repository, for example `$HOME/ncs/nrf-softsim-dev`.

#### Note

> SoftSIM is relying on some default data in the storage partition. This section of the flash can generated and flashed manually (see steps below) or automatically included by `CONFIG_SOFTSIM_BUNDLE_TEMPLATE_HEX=y` - This is generally recommended. 

Reviewer: ```suggestion
SoftSIM is relying on some default data in the storage partition. This section of the flash can be generated and flashed manually (see steps below) or, as we recommend, automatically included by `CONFIG_SOFTSIM_BUNDLE_TEMPLATE_HEX=y` 
```

If the partition table of the application changes, for example due to another partition changing size, the template profile must be rebuilt and flashed again.
The partition table can be checked at any time with:
```
west build -t partition_manager_report
```

#### Note

Some applications will fail to link with error `zephyr/zephyr_pre0.elf uses VFP register arguments` (for example `modem_shell`). In this case it is required to also enable `CONFIG_FP_SOFTABI=y`. It is suggested to create an additional Kconfig overlay for application specific SoftSIM configurations and add them to `overlay-softsim.conf` inside
the application directory.
The application can then be built like this:
```
west build -b nrf9160dk_nrf9160_ns -- "-DOVERLAY_CONFIG=$PATH_TO_ONOMONDO_SOFTSIM/overlay-softsim.conf;overlay-softsim.conf"
```

#### Note

For very large applications, it is required to disable features in order to reduce the size of the application binary and leave space on Flash for the SIM profile. 
During the build step, a Flash overflow error will be reported if this requirement is not satisfied.
The same principle applies for RAM requirements.

#### Note

> Onomondo SoftSIM uses the heap memory pool. It is expected that `CONFIG_HEAP_MEM_POOL_SIZE` is at least `30000` so if the target application also uses the heap please

Reviewer: ```suggestion
Onomondo SoftSIM uses the heap memory pool. It is expected that `CONFIG_HEAP_MEM_POOL_SIZE` is at least `30000`, so if the target application also uses the heap, please
```

To accomodate that we've created a bootstrapping filesystem that should be flashed together with the application.

<p align="center">
 <img width="338" src="https://github.com/onomondo/nrf-softsim/assets/46489969/e77404ed-f8fd-46c8-98d8-054258727b8b">
</p>

The list of files is fairly involved - but in the end only a subset of files are ever accessed. 

> Internally this is a NVS partition which is a `key-value` store type. It is pretty compact and _good enough_ for what we need it for. So the previous request of reading the `ICCID` internally translates to something like:

Reviewer: ```suggestion
Internally this is a NVS partition which is a `key-value` store type. It is pretty compact and generally sufficient. The previous request of reading the `ICCID` internally translates to something similar to:
```

`00a408040000022fe20168` -> `open('/3f00/2fe2)` -> `nvs_read(id=14)`

> We've made a caching layer as well to avoid i) slow reads ii) excessive writes to flash. So actually the SoftSIM profile data looks something like this:

Reviewer: ```suggestion
We've made a caching layer as well to avoid i) slow reads and ii) excessive writes to flash. So, the SoftSIM profile data looks something like:
```

<p align="center">
 <img width="358" src="https://github.com/onomondo/nrf-softsim/assets/46489969/c03113b3-f41b-41c7-b681-0e2b09f7ee7b">
</p>

> The first entry is used to translate between paths (`'3f00/2fe2`) to an actual `NVS key`. It contains an ordered list of files sorted by frequency of access - i.e. the 'master file, 3f00' is in the top since it is most frequenctly accessed. 

Reviewer: ```suggestion
The first entry is used to translate between paths (`'3f00/2fe2`) to an actual `NVS key`. It contains an ordered list of files sorted by frequency of access - i.e. the 'master file, 3f00' is in the top, since it is most frequently accessed. 
```

> The lsit is read and parsed to a linked list - and this makes the base for all cached operations. The order makes the lookup very fast in most cases.

Reviewer: ```suggestion
The list is read and parsed to a linked list - and this makes the base for all cached operations. The order makes the lookup very fast.
```

> When a device finds a network it want to attach to something like this happens:

Reviewer: ```suggestion
When a device finds a network it wants to attach to, something like this happens (simplified version):
```

<p align="center">
<img height="600" src="https://github.com/onomondo/nrf-softsim/assets/46489969/19083bea-2727-48d6-ad50-63f80384e4d8">
</p>

> That is greatly simplified. Step 5 is a SIM command as well `AUTHENTICATE EVEN` and it is running the `milenage algorithm` that also creates session keys etc (and checks that the network in fact isn't an imposter). 

Reviewer: ```suggestion
Step 5 is a SIM command as well `AUTHENTICATE EVEN` and it is running the `milenage algorithm` that also creates session keys, checks that the network in fact isn't an imposter, etc.  
```

CONFIG_FLASH_PAGE_LAYOUT=y
CONFIG_MPU_ALLOW_FLASH_WRITE=y

# Enable nordic security backend and PSA APIs
# probably already the default? 
CONFIG_NRF_SECURITY=y
CONFIG_PSA_CRYPTO_DRIVER_CC3XX=y
CONFIG_PSA_WANT_ALG_CBC_NO_PADDING=y
CONFIG_PSA_WANT_ALG_CMAC=y
CONFIG_PSA_WANT_KEY_TYPE_AES=y

CONFIG_NVS=y
CONFIG_PM_PARTITION_SIZE_NVS_STORAGE=0x8000 

# include softsim in build
CONFIG_SOFTSIM=y
CONFIG_SOFTSIM_AUTO_INIT=y
CONFIG_SOFTSIM_BUNDLE_TEMPLATE_HEX=y

# Currently needed for persistent key storage
CONFIG_TFM_PROFILE_TYPE_NOT_SET=y
```

# Flashing your device:
### By command line:

`nrfjprog -f NRF91 --sectorerase --log --program /path/to/build/zephyr/merged.hex --reset`

# Some more details:
#### Heap

> `mem.h`/`heap_port.c` can now optionally be used to implement custom allocators - i.e. `k_malloc()` to avoid conflict with stdc heap pools. We default to `k_malloc()/k_free()`

Reviewer: ```suggestion
`mem.h`/`heap_port.c` can now be optionally used to implement custom allocators - i.e. `k_malloc()` to avoid conflict with stdc heap pools. We default to `k_malloc()/k_free()`
```
